### PR TITLE
feat: disable file:// urls when hardening enabled

### PR DESCRIPTION
**Pull request by gwossum**
Stacks and templates allow specifying file:// URLs. Add option to disable their use for people who don't require them.

### CODE REVIEW

File: pkger/parser.go [hunk 1]
```diff
@@ -102,8 +103,71 @@ func Parse(encoding Encoding, readerFn ReaderFn, opts ...ValidateOptFn) (*Templa
...
+	}
+	size = int(size64)
+
+	// Read file
+	data := make([]byte, size)
+	b, err := f.Read(data)
+	if err != nil {
+		return nil, err
+	} else if b != size {
+		return nil, errors.New("short read")
```
[the remark this answers is not in the document]
Comment (davidby-influx) [reply pkger/parser.go:163]: It was in ioutil, so maybe deprecated or gone?

**Review by davidby-influx — CHANGES_REQUESTED** [review]
A first pass; I can do a deeper review another day.  Mostly about file names in errors...

File: pkg/fs/special_linux.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+	getDevId := func(st fs.FileInfo) (uint64, error) {
+		st_sys_any := st.Sys()
+		if st_sys_any == nil {
+			return 0, errors.New("nil returned by fs.FileInfo.Sys")
+		}
+
+		st_sys, ok := st_sys_any.(*syscall.Stat_t)
+		if !ok {
+			return 0, errors.New("could not convert st.sys() to a *syscall.Stat_t")
```
[the remark this answers is not in the document]
Comment (gwossum) [reply pkg/fs/special_linux.go:29]: I'm going add the filename to the caller's error. `IsSepcialFSFromFileInfo` does not have access to the full path of the filename, only the base path.

File: pkger/parser.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gwossum) [reply pkger/parser.go:163]: There is an `ErrShortWrite`, but no `ErrShortRead` that I know of.

[the remark this answers is not in the document]
Comment (davidby-influx) [reply pkger/parser.go]: Found it; it's in kafka and sqlite packages, not appropriate here. Sorry for the misdirection.

File: pkger/parser.go [hunk 3]
```diff
@@ -102,8 +104,71 @@ func Parse(encoding Encoding, readerFn ReaderFn, opts ...ValidateOptFn) (*Templa
...
+
+	// only support reading regular files
+	if st.Mode()&os.ModeType != 0 {
+		return nil, fmt.Errorf("not a regular file: %q", name)
+	}
+
+	// limit how much we read into RAM
+	var size int
+	size64 := st.Size()
+	if size64 > limitReadFileMaxSize {
```
Comment (davidby-influx) [inline pkger/parser.go:149]: Do we want to add a `&& limitReadFileMaxSize > 0` in case it becomes a variable some day?  Not blocking on merge, just a thought.

**Review by davidby-influx — APPROVED** [review]
LGTM. One question, not blocking.

**Review by jdstrand — CHANGES_REQUESTED** [review]
Thanks for the PR! Most of my in-line comments are observations, but there are a few places where it would be good to update comments (eg, remocal and zero size files in /proc).

I also did some manual testing (using the amd64 deb from https://app.circleci.com/pipelines/github/influxdata/influxdb/38891/workflows/ec8d697d-ee31-46a3-9f15-2f1238f2eb8b/jobs/363718/artifacts) for /api/v2/stacks:
```
$ cat ~/stuff/ok.json
[
   {
      "apiVersion": "influxdata.com/v2alpha1",
      "kind": "Label",
      "metadata": {
         "name": "label-1"
      },
      "spec": {
         "color": "#eee888",
         "description": "desc_1"
      }
   },
   {
      "apiVersion": "influxdata.com/v2alpha1",
      "kind": "Bucket",
      "metadata": {
         "name": "rucket-1"
      },
      "spec": {
         "associations": [
            {
               "kind": "Label",
               "name": "label-1"
            }
         ],
         "description": "desc_1",
         "retentionRules": [
            {
               "everySeconds": 10000,
               "type": "expire"
            }
         ]
      }
   }
]

# in one terminal, start a http server to serve this over http://
$ cd ~/stuff
$ python3 -m http.server

# via http://
$ curl --insecure -H "Authorization: Token $TOKEN" --request POST "$URL/api/v2/stacks" --data "{\"orgID\":\"$ORGID\", \"name\":\"some name\", \"description\": \"some desc\", \"urls\": [\"http://localhost:8000/ok.json\"]}"
{
	"id": "0ce72523377b4000",
	"orgID": "697959a536e728a2",
	"createdAt": "2024-04-17T10:05:13.181224325-05:00",
...
$ influx apply -o $ORG --stack-id 0ce72523377b4000
LABELS    +add | -remove | unchanged
+-----+---------------+------------------+---------------+---------+-------------+
| +/- | METADATA NAME |        ID        | RESOURCE NAME |  COLOR  | DESCRIPTION |
+-----+---------------+------------------+---------------+---------+-------------+
| +   | label-1       | 0000000000000000 | label-1       | #eee888 | desc_1      |
+-----+---------------+------------------+---------------+---------+-------------+
|                                                           TOTAL  |      1      |
+-----+---------------+------------------+---------------+---------+-------------+
...

$ influx stacks rm --stack-id 0ce72523377b4000

# now try with file://
$ curl --insecure -H "Authorization: Token $TOKEN" --request POST "$URL/api/v2/stacks" --data "{\"orgID\":\"$ORGID\", \"name\":\"some name\", \"description\": \"some desc\", \"urls\": [\"file:///home/ubuntu/stuff/ok.json\"]}"
{
	"id": "0ce725ef413b4000",
	"orgID": "697959a536e728a2",
	"createdAt": "2024-04-17T10:08:42.116599762-05:00",
...

$ influx apply -o $ORG --stack-id 0ce725ef413b4000
LABELS    +add | -remove | unchanged
+-----+---------------+------------------+---------------+---------+-------------+
| +/- | METADATA NAME |        ID        | RESOURCE NAME |  COLOR  | DESCRIPTION |
+-----+---------------+------------------+---------------+---------+-------------+
| +   | label-1       | 0000000000000000 | label-1       | #eee888 | desc_1      |
+-----+---------------+------------------+---------------+---------+-------------+
|                                                           TOTAL  |      1      |
+-----+---------------+------------------+---------------+---------+-------------+
...

$ influx stacks rm --stack-id 0ce725ef413b4000
```

Now try stacks with 'hardening-enabled = true':
```
# the curl is expected to pass
$ curl --insecure -H "Authorization: Token $TOKEN" --request POST "$URL/api/v2/stacks" --data "{\"orgID\":\"$ORGID\", \"name\":\"some name\", \"description\": \"some desc\", \"urls\": [\"file:///home/ubuntu/stuff/ok.json\"]}"
{
	"id": "0ce72690ee408000",
	"orgID": "697959a536e728a2",
	"createdAt": "2024-04-17T10:11:27.673418946-05:00",
...

# the apply is not
$ influx apply -o $ORG --stack-id 0ce72690ee408000
Error: failed to check template impact: 400 Bad Request: invalid URL scheme: invalid URL scheme
$ influx stacks rm --stack-id 0ce72690ee408000
```

Now try stacks with just 'template-file-urls-disabled = true':
```
# the curl is expected to pass
$ curl --insecure -H "Authorization: Token $TOKEN" --request POST "$URL/api/v2/stacks" --data "{\"orgID\":\"$ORGID\", \"name\":\"some name\", \"description\": \"some desc\", \"urls\": [\"file:///home/ubuntu/stuff/ok.json\"]}"
{
	"id": "0ce7271e9d86b000",
	"orgID": "697959a536e728a2",
	"createdAt": "2024-04-17T10:11:27.673418946-05:00",
...

# the apply is not
$ influx apply -o $ORG --stack-id 0ce7271e9d86b000
Error: failed to check template impact: 400 Bad Request: invalid URL scheme: invalid URL scheme
$ influx stacks rm --stack-id 0ce7271e9d86b000
```


Now try /api/v2/templates/apply directly with neither hardening-enabled nor template-file-urls-disabled set:
```
# test http:// still works with /api/v2/templates/apply
$ curl --insecure -H "Authorization: Token $TOKEN" -H "Content-Type: application/json" --request POST "$URL/api/v2/templates/apply" --data "{\"orgID\": \"$ORGID\", \"dryRun\": true, \"remotes\": [{\"url\": \"http://localhost:8000/ok.json\"}]}"
{
	"sources": [
		"http://localhost:8000/ok.json"
	],
	"stackID": "",
...

# influx apply --encoding json -f works too  # the -f path is on the client, not server
$ influx apply --encoding json -f ~/stuff/ok.json
LABELS    +add | -remove | unchanged
+-----+---------------+------------------+---------------+---------+-------------+
| +/- | METADATA NAME |        ID        | RESOURCE NAME |  COLOR  | DESCRIPTION |
+-----+---------------+------------------+---------------+---------+-------------+
| +   | label-1       | 0000000000000000 | label-1       | #eee888 | desc_1      |
+-----+---------------+------------------+---------------+---------+-------------+
|                                                           TOTAL  |      1      |
+-----+---------------+------------------+---------------+---------+-------------+
...

# now try with file:///home/ubuntu/stuff/ok.json 
# Curiously, I was thinking should've been allowed since neither hardening-enabled nor
# template-file-urls-disabled was specified in the config file
$ curl --insecure -H "Authorization: Token $TOKEN" -H "Content-Type: application/json" --request POST "$URL/api/v2/templates/apply" --data "{\"orgID\": \"$ORGID\", \"dryRun\": true, \"remotes\": [{\"url\": \"file:///home/ubuntu/stuff/ok.json\"}]}"
{
	"code": "unprocessable entity",
	"message": "template from url[file:///home/ubuntu/stuff/ok.json] had an issue: Get \"file:///home/ubuntu/stuff/ok.json\": unsupported protocol scheme \"file\""
```

I then tried with the official 2.7.6-1 amd64 deb, and it also fails to process `file:///home/ubuntu/stuff/ok.json` with /api/v2/templates/apply. I then tried 2.1.1 (which predates https://github.com/influxdata/influxdb/pull/23030, which I though might've changed the behavior) and it fails in the same way.

So, /api/v2/stacks is working as designed by the PR. /api/v2/templates/apply unconditionally denies use of file:/// URLs and has for years. The difference in behavior is odd and makes me want to say that we should unconditionally disallow file:/// URLs in both for consistency, but we know that file:/// URLs are used with stacks, so I don't think we can do that.

Long story short, this PR is functionally correct.

File: cmd/influxd/launcher/cmd.go [hunk 4]
```diff
@@ -657,7 +661,16 @@ func (o *InfluxdOpts) BindCliOpts() []cli.Opt {
 			DestP:   &o.HardeningEnabled,
 			Flag:    "hardening-enabled",
 			Default: o.HardeningEnabled,
-			Desc:    "enable hardening options (disallow private IPs within flux and templates HTTP requests)",
+			Desc:    "enable hardening options (disallow private IPs within flux and templates HTTP requests, template file URLs)",
```
[the remark this answers is not in the document]
Comment (gwossum) [reply cmd/influxd/launcher/cmd.go:664]: I was trying to keep the line length short, but it looks like other help lines are already long. I'll make this change.

File: pkg/fs/special_linux.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		return false, nil
+	}
+
+	// We know the file is in a special file system, but we'll make an
+	// exception for tmpfs, which some distros use for /tmp.
+	// Since the minor IDs are assigned dynamically, we'll find the device ID
+	// for /tmp. If /tmp's device ID matches this st's, then it is safe to assume
+	// the file is in tmpfs. Otherwise, it is a special device that is not tmpfs.
+	// Note that if /tmp is not tmpfs, then the device IDs won't match since
+	// tmp would be a standard file system.
```
[the remark this answers is not in the document]
Comment (gwossum) [reply pkg/fs/special_linux.go:49]: Yeah, I was trying to get that point across that if /tmp (or $TMPDIR) is a standard file system, you won't even get here. I can clarify it some.

File: pkg/fs/special_linux.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		// say the file is in tmpfs. We'll still go ahead and call it a special file.
+		return true, nil
+	}
+	tmpDevId, err := getDevId(tmpSt)
+	if err != nil {
+		// See above for why we're returning an error here.
+		return true, nil
+	}
+	// It's a special file unless the device ID matches /tmp's ID.
+	return devId != tmpDevId, nil
```
[the remark this answers is not in the document]
Comment (gwossum) [reply pkg/fs/special_linux.go:62]: IMHO, `/dev/shm` is only for POSIX shard memory, and it's not specifically in FHS. `/run` is definitely interesting, though, because it is almost always tmpfs and is guaranteed by FHS. It's also what systemd says you should use instead of `/dev/shm`. I'm thinking we should have an additional special case check for `/run`. We could also add a check against `/dev/shm` if you want.

File: pkg/fs/special_linux.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	"os"
+	"syscall"
+
+	"golang.org/x/sys/unix"
+)
+
+// IsSpecialFSFromFileInfo determines if a file resides on a special file
+// system (e.g. /proc, /dev/, /sys) based on its fs.FileInfo.
+// The bool return value should be ignored if err is not nil.
+func IsSpecialFSFromFileInfo(st fs.FileInfo) (bool, error) {
```
[the remark this answers is not in the document]
Comment (gwossum) [reply pkg/fs/special_linux.go:15]: Longer term, probably yes. However, I don't yet know how to detect special file systems on Darwin. When I looked into it, I also discovered Darwin / MacOS doesn't use special file systems to the extent Linux does. For instance, there is a `/proc` file system but it is not mounted by default.

File: pkger/service.go [hunk 8]
```diff
@@ -3101,11 +3111,35 @@ func (s *Service) getStackRemoteTemplates(ctx context.Context, stackID platform.
...
+
+			if s.fileUrlsDisabled {
+				return nil, &errors2.Error{
+					Code: errors2.EInvalid,
+					Msg:  "invalid URL scheme",
+					Err:  errors.New("invalid URL scheme"),
+				}
+			}
+
+			readerFn = FromFile(u.Path, true)
```
[the remark this answers is not in the document]
Comment (gwossum) [reply pkger/service.go:3126]: I tried this out, but it broke a lot of tests. Some of the test cases require turning off the checks, and making these tests work with the checks seemed complicated.

File: pkg/fs/special_linux.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jdstrand) [reply pkg/fs/special_linux.go:62]: I know some people have gotten /dev/shm (or /run/shm) under their fingers for having a reliably memory-backed sticky dir (/tmp-style) on Linux (eg, to generate private keys to keep on disk). That said, it feels unlikely that someone would put a template in /dev/shm (or /run/shm on a server) to then start running stacks command for influxdb. In thinking about this more, at most this is a release note IMHO.

As for `/run`, it's not impossible to imagine that people might put things under `/run/user` or `/run/something`. influxdb doesn't have a /run/influxdb directory with 2.x and /run/user is going to be challenging to use for users since it is only accessible to the user (and not the influxdb user) so I don't think this is required.

**Review by jdstrand — CHANGES_REQUESTED** [review]
This is probably fine as is, but I have a small pedantic suggestion for `/dev/shm` and a few comment update suggestions. I'm not sure if the tests are flaky (I didn't check), but fyi, not all the tests are passing.

File: pkg/fs/special_linux.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	// Since the minor IDs are assigned dynamically, we'll find the device ID
+	// for each common tmpfs mount point, If the mount point's device ID matches this st's,
+	// then it is reasonable to assume the file is in tmpfs. If the device ID
+	// does not match, then st is not located in that special file system so we
+	// can't give an exception based on that file system root. This check is still
+	// valid even if the directory we check against isn't mounted as tmpfs, because
+	// the device ID won't match so we won't grant a tmpfs exception based on it.
+	// On Linux, every tmpfs mount has a different device ID, so we need to check
+	// against all common ones that might be in use.
+	tmpfsMounts := []string{"/tmp", "/run", "/dev/shm"}
```
[the remark this answers is not in the document]
Comment (gwossum) [reply pkg/fs/special_linux.go:55]: The good news is that the check for the special filesystem and tmpfs exceptions is against the target of the symlink, so the same code should work for that situation. I've added test cases with symlinks as well.

File: pkg/fs/special_linux.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gwossum) [reply pkg/fs/special_linux.go]: Updated code should handle `/run` and `/run/user/$(id -u)` properly. Handling of `/run/user/$(id -u)` relies on the setting of `XDG_RUNTIME_DIR` environment variable.

File: pkg/fs/special_linux.go [hunk 10]
```diff
@@ -41,23 +42,45 @@ func IsSpecialFSFromFileInfo(st fs.FileInfo) (bool, error) {
...
+	}
+	getFileDevId := func(n string) (uint64, error) {
+		fSt, err := os.Stat(n)
+		if err != nil {
+			return math.MaxUint64, err
+		}
+		fDevId, err := getDevId(fSt)
+		if err != nil {
+			// See above for why we're returning an error here.
+			return math.MaxUint64, nil
```
[the remark this answers is not in the document]
Comment (gwossum) [reply pkg/fs/special_linux.go]: Change code so that error is returned for improved error reporting.

File: pkg/fs/special_linux.go [hunk 11]
```diff
@@ -41,23 +42,45 @@ func IsSpecialFSFromFileInfo(st fs.FileInfo) (bool, error) {
...
+		}
+		return fDevId, nil
+	}
+	for _, fn := range tmpfsMounts {
+		// We ignore errors if getFileDevId fails, which could
+		// mean that the file (e.g. /run) doesn't exist. The error
```
[the remark this answers is not in the document]
Comment (gwossum) [reply pkg/fs/special_linux.go]: Complete redid comment.

File: pkg/fs/special_linux.go [hunk 12]
```diff
@@ -41,23 +42,45 @@ func IsSpecialFSFromFileInfo(st fs.FileInfo) (bool, error) {
...
-	// for /tmp. If /tmp's device ID matches this st's, then it is safe to assume
-	// the file is in tmpfs. Otherwise, it is a special device that is not tmpfs.
-	// Note that if /tmp is not tmpfs, then the device IDs won't match since
-	// tmp would be a standard file system.
-	tmpSt, err := os.Stat(os.TempDir())
-	if err != nil {
-		// We got an error getting stats on /tmp, but that just means we can't
-		// say the file is in tmpfs. We'll still go ahead and call it a special file.
-		return true, nil
+	// for each common tmpfs mount point, If the mount point's device ID matches this st's,
```
[the remark this answers is not in the document]
Comment (gwossum) [reply pkg/fs/special_linux.go]: Fixed.

File: pkger/parser.go (hunk 3 above)
Comment (gwossum) [reply pkger/parser.go]: Added in latest commit.

File: pkger/parser_test.go [hunk 13]
```diff
@@ -4825,6 +4829,249 @@ func Test_validGeometry(t *testing.T) {
...
+			candidateDirs := []string{
+				"/dev/shm",
+				"/run",
+				"/tmp",
+				os.Getenv("XDG_RUNTIME_DIR"),
+				os.TempDir(),
+			}
+			var actualDirs []string
+			for _, dir := range candidateDirs {
+				if dir == "" {
```
Comment (jdstrand) [inline pkger/parser_test.go:4983]: Nitpick: either here or the `os.Getenv("XDG_RUNTIME_DIR")` might benefit from a comment that this is for the other. Not a blocker.

**Review by jdstrand — APPROVED** [review]
Thanks for the updates! LGTM. I like how `IsSpecialFSFromFileInfo()` fails closed. One teensy nitpick in-line that you can feel free to ignore.
